src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -79,6 +88,49 @@ public synchronized void addTask(Task person) {
         indicateTaskBookChanged();
     }
 
+    //=============for undo and redo===================================
+    @Override
+    public Command undo() throws IllegalValueException {
+
+        if (head <= 0) {
+            System.out.println(head);
+            throw new IllegalValueException("No actions to undo");
+        }
+        head--;
+        Commit commit = commits.get(head);
+        resetData(new TaskBook(commit.getTaskBook()));
+        return commit.getCommand();
+    }
+
+    @Override
+    public void recordState(Command command) {
+        commits.add(new Commit(command, new TaskBook(getTaskBook())));
+        this.head = commits.size() - 1;
```

Comment:
undo is a command, it should be tested with other commands and not recordState()